File: ios/Yizhidao/Data/SavedAIAnalysis.swift
import Foundation

func aiAdviceDisplayItems(advice: [String], risks: [String]) -> [String] {
    let parts = risks.compactMap { raw -> String? in
        var text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return nil }
        for prefix in ["须防：", "须防:", "須防：", "須防:"] where text.hasPrefix(prefix) {
            text = String(text.dropFirst(prefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
            break
        }
        return text.isEmpty ? nil : text
    }
    guard !parts.isEmpty else { return advice }
    return advice + ["须防：\(parts.joined(separator: "；"))"]
}

struct SavedAIFollowUp: Codable, Hashable, Identifiable {
    let id: UUID
    let user: String
    let assistant: String
    let advice: [String]
    let askNext: [String]

    enum CodingKeys: String, CodingKey {
        case id, user, assistant, advice, askNext
    }

    init(
        id: UUID = UUID(),
        user: String,
        assistant: String,
        advice: [String] = [],
        askNext: [String] = []
    ) {
        self.id = id
        self.user = user
        self.assistant = assistant
        self.advice = advice
        self.askNext = askNext
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        user = try container.decode(String.self, forKey: .user)
        assistant = try container.decode(String.self, forKey: .assistant)
        advice = try container.decodeIfPresent([String].self, forKey: .advice) ?? []
        askNext = try container.decodeIfPresent([String].self, forKey: .askNext) ?? []
    }
}

struct SavedAIContent: Codable, Hashable {
    var summary: String
    var focus: String
    var advice: [String]
    var direction: String
    var risks: [String]
    var askNext: [String]

    enum CodingKeys: String, CodingKey {
        case summary, focus, advice, direction, risks, askNext
    }

    init(
        summary: String,
        focus: String,
        advice: [String],
        direction: String = "",
        risks: [String] = [],
        askNext: [String] = []
    ) {
        self.summary = summary
        self.focus = focus
        self.advice = advice
        self.direction = direction
        self.risks = risks
        self.askNext = askNext
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        summary = try container.decode(String.self, forKey: .summary)
        focus = try container.decode(String.self, forKey: .focus)
        advice = try container.decode([String].self, forKey: .advice)
        direction = try container.decodeIfPresent(String.self, forKey: .direction) ?? ""
        risks = try container.decodeIfPresent([String].self, forKey: .risks) ?? []
        askNext = try container.decodeIfPresent([String].self, forKey: .askNext) ?? []
    }
}

struct SavedAIAnalysis: Codable, Identifiable, Hashable {
    let id: UUID
    let createdAt: Date
    var updatedAt: Date
    var readingRecordID: UUID?
    let methodRaw: String
    let question: String?
    let primaryNumber: Int
    let resultingNumber: Int?
    let lines: [Int]
    let movingPositions: [Int]
    var analysis: SavedAIContent
    var followUps: [SavedAIFollowUp]

    var method: CastingMethod {
        CastingMethod(rawValue: methodRaw) ?? .digitalManual
    }

    var fingerprint: String {
        Self.fingerprint(
            methodRaw: methodRaw,
            createdAt: createdAt,
            primaryNumber: primaryNumber,
            lines: lines
        )
    }

    static func fingerprint(
        methodRaw: String,
        createdAt: Date,
        primaryNumber: Int,
        lines: [Int]
    ) -> String {
        "\(methodRaw)|\(createdAt.timeIntervalSince1970)|\(primaryNumber)|\(lines.map(String.init).joined(separator: ","))"
    }

    static func fingerprint(result: CastResult) -> String {
        fingerprint(
            methodRaw: result.method.rawValue,
            createdAt: result.createdAt,
            primaryNumber: result.primaryNumber,
            lines: result.lines.map(\.rawValue)
        )
    }

    func toCastResult() -> CastResult {
        CastResult(
            method: method,
            createdAt: createdAt,
            question: question,
            numbers: nil,
            primaryNumber: primaryNumber,
            resultingNumber: resultingNumber,
            lines: lines.compactMap(LineValue.init(rawValue:)),
            movingPositions: movingPositions
        )
    }

    static func make(
        result: CastResult,
        analysis: SavedAIContent,
        followUps: [SavedAIFollowUp],
        readingRecordID: UUID? = nil,
        existingID: UUID? = nil
    ) -> SavedAIAnalysis {
        SavedAIAnalysis(
            id: existingID ?? UUID(),
            createdAt: result.createdAt,
            updatedAt: Date(),
            readingRecordID: readingRecordID,
            methodRaw: result.method.rawValue,
            question: result.question,
            primaryNumber: result.primaryNumber,
            resultingNumber: result.resultingNumber,
            lines: result.lines.map(\.rawValue),
            movingPositions: result.movingPositions,
            analysis: analysis,
            followUps: followUps
        )
    }

    enum CodingKeys: String, CodingKey {
        case id, createdAt, updatedAt, readingRecordID, methodRaw, question
        case primaryNumber, resultingNumber, lines, movingPositions, analysis, followUps
    }

    init(
        id: UUID,
        createdAt: Date,
        updatedAt: Date,
        readingRecordID: UUID? = nil,
        methodRaw: String,
        question: String?,
        primaryNumber: Int,
        resultingNumber: Int?,
        lines: [Int],
        movingPositions: [Int],
        analysis: SavedAIContent,
        followUps: [SavedAIFollowUp]
    ) {
        self.id = id
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.readingRecordID = readingRecordID
        self.methodRaw = methodRaw
        self.question = question
        self.primaryNumber = primaryNumber
        self.resultingNumber = resultingNumber
        self.lines = lines
        self.movingPositions = movingPositions
        self.analysis = analysis
        self.followUps = followUps
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        updatedAt = try container.decode(Date.self, forKey: .updatedAt)
        readingRecordID = try container.decodeIfPresent(UUID.self, forKey: .readingRecordID)
        methodRaw = try container.decode(String.self, forKey: .methodRaw)
        question = try container.decodeIfPresent(String.self, forKey: .question)
        primaryNumber = try container.decode(Int.self, forKey: .primaryNumber)
        resultingNumber = try container.decodeIfPresent(Int.self, forKey: .resultingNumber)
        lines = try container.decode([Int].self, forKey: .lines)
        movingPositions = try container.decode([Int].self, forKey: .movingPositions)
        analysis = try container.decode(SavedAIContent.self, forKey: .analysis)
        followUps = try container.decodeIfPresent([SavedAIFollowUp].self, forKey: .followUps) ?? []
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(createdAt, forKey: .createdAt)
        try container.encode(updatedAt, forKey: .updatedAt)
        try container.encodeIfPresent(readingRecordID, forKey: .readingRecordID)
        try container.encode(methodRaw, forKey: .methodRaw)
        try container.encodeIfPresent(question, forKey: .question)
        try container.encode(primaryNumber, forKey: .primaryNumber)
        try container.encodeIfPresent(resultingNumber, forKey: .resultingNumber)
        try container.encode(lines, forKey: .lines)
        try container.encode(movingPositions, forKey: .movingPositions)
        try container.encode(analysis, forKey: .analysis)
        try container.encode(followUps, forKey: .followUps)
    }
}

extension Notification.Name {
    static let savedAIAnalysesDidChange = Notification.Name("yizhidao.savedAIAnalysesDidChange")
}

enum SavedAIAnalysisStore {
    private static let key = "ai.saved.analyses.v1"

    static func load() -> [SavedAIAnalysis] {
        guard let data = UserDefaults.standard.data(forKey: key),
              let items = try? JSONDecoder().decode([SavedAIAnalysis].self, from: data)
        else { return [] }
        return items.sorted { $0.updatedAt > $1.updatedAt }
    }

    static func save(_ items: [SavedAIAnalysis]) {
        guard let data = try? JSONEncoder().encode(items) else { return }
        UserDefaults.standard.set(data, forKey: key)
        NotificationCenter.default.post(name: .savedAIAnalysesDidChange, object: nil)
    }

    static func upsert(_ item: SavedAIAnalysis) {
        var items = load()
        if let recordID = item.readingRecordID,
           let index = items.firstIndex(where: { $0.readingRecordID == recordID }) {
            var merged = item
            if items[index].id != item.id {
                merged = SavedAIAnalysis.make(
                    result: item.toCastResult(),
                    analysis: item.analysis,
                    followUps: item.followUps,
                    readingRecordID: recordID,
                    existingID: items[index].id
                )
            }
            items[index] = merged
        } else if let index = items.firstIndex(where: { $0.id == item.id }) {
            items[index] = item
        } else {
            items.insert(item, at: 0)
        }
        save(items)
    }

    static func find(recordID: UUID?, result: CastResult) -> SavedAIAnalysis? {
        let items = load()
        if let recordID, let hit = items.first(where: { $0.readingRecordID == recordID }) {
            return hit
        }
        let fp = SavedAIAnalysis.fingerprint(result: result)
        return items.first { $0.fingerprint == fp }
    }

    static func remove(id: UUID) {
        var items = load()
        items.removeAll { $0.id == id }
        save(items)
    }
}

File: android/app/src/main/java/com/yizhidao/app/ai/SavedAIAnalysis.kt
package com.yizhidao.app.ai

import android.content.Context
import com.yizhidao.CastResult
import com.yizhidao.CastingMethod
import com.yizhidao.LineValue
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.time.Instant
import java.util.UUID

fun aiAdviceDisplayItems(advice: List<String>, risks: List<String> = emptyList()): List<String> {
    val prefixes = listOf("须防：", "须防:", "須防：", "須防:")
    val parts = risks.mapNotNull { raw ->
        var text = raw.trim()
        if (text.isEmpty()) return@mapNotNull null
        val prefix = prefixes.firstOrNull { text.startsWith(it) }
        if (prefix != null) text = text.removePrefix(prefix).trim()
        text.ifEmpty { null }
    }
    if (parts.isEmpty()) return advice
    return advice + listOf("须防：${parts.joinToString("；")}")
}

@Serializable
data class SavedAIFollowUp(
    val id: String = UUID.randomUUID().toString(),
    val user: String,
    val assistant: String,
    val advice: List<String> = emptyList(),
    val askNext: List<String> = emptyList(),
)

@Serializable
data class SavedAIContent(
    val summary: String,
    val focus: String,
    val advice: List<String>,
    val direction: String = "",
    val risks: List<String> = emptyList(),
    val askNext: List<String> = emptyList(),
)

@Serializable
data class SavedAIAnalysis(
    val id: String = UUID.randomUUID().toString(),
    val createdAtEpochMs: Long,
    val updatedAtEpochMs: Long,
    val readingRecordId: String? = null,
    val methodRaw: String,
    val question: String? = null,
    val primaryNumber: Int,
    val resultingNumber: Int? = null,
    val lines: List<Int>,
    val movingPositions: List<Int>,
    val analysis: SavedAIContent,
    val followUps: List<SavedAIFollowUp> = emptyList(),
) {
    val method: CastingMethod get() = CastingMethod.fromRaw(methodRaw)
    val createdAt: Instant get() = Instant.ofEpochMilli(createdAtEpochMs)
    val updatedAt: Instant get() = Instant.ofEpochMilli(updatedAtEpochMs)

    fun fingerprint(): String = fingerprint(methodRaw, createdAtEpochMs, primaryNumber, lines)

    fun toCastResult(): CastResult = CastResult(
        method = method,
        createdAt = createdAt,
        question = question,
        numbers = null,
        primaryNumber = primaryNumber,
        resultingNumber = resultingNumber,
        lines = lines.mapNotNull(LineValue::fromRaw),
        movingPositions = movingPositions,
    )

    companion object {
        fun fingerprint(
            methodRaw: String,
            createdAtEpochMs: Long,
            primaryNumber: Int,
            lines: List<Int>,
        ): String = "$methodRaw|$createdAtEpochMs|$primaryNumber|${lines.joinToString(",")}"

        fun fingerprint(result: CastResult): String = fingerprint(
            methodRaw = result.method.raw,
            createdAtEpochMs = result.createdAt.toEpochMilli(),
            primaryNumber = result.primaryNumber,
            lines = result.lines.map { it.rawValue },
        )

        fun make(
            result: CastResult,
            analysis: SavedAIContent,
            followUps: List<SavedAIFollowUp>,
            readingRecordId: String? = null,
            existingId: String? = null,
        ): SavedAIAnalysis {
            val now = System.currentTimeMillis()
            return SavedAIAnalysis(
                id = existingId ?: UUID.randomUUID().toString(),
                createdAtEpochMs = result.createdAt.toEpochMilli(),
                updatedAtEpochMs = now,
                readingRecordId = readingRecordId,
                methodRaw = result.method.raw,
                question = result.question,
                primaryNumber = result.primaryNumber,
                resultingNumber = result.resultingNumber,
                lines = result.lines.map { it.rawValue },
                movingPositions = result.movingPositions,
                analysis = analysis,
                followUps = followUps,
            )
        }
    }
}

class SavedAIAnalysisStore(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }
    private val _items = MutableStateFlow(load())
    val items: StateFlow<List<SavedAIAnalysis>> = _items.asStateFlow()

    fun load(): List<SavedAIAnalysis> {
        val raw = prefs.getString(KEY, null) ?: return emptyList()
        return runCatching { json.decodeFromString<List<SavedAIAnalysis>>(raw) }
            .getOrDefault(emptyList())
            .sortedByDescending { it.updatedAtEpochMs }
    }

    fun upsert(item: SavedAIAnalysis) {
        val current = _items.value.toMutableList()
        val byRecord = item.readingRecordId
            ?.takeIf { it.isNotBlank() }
            ?.let { rid -> current.indexOfFirst { it.readingRecordId == rid } }
            ?: -1
        val byId = current.indexOfFirst { it.id == item.id }
        val byFingerprint = current.indexOfFirst { it.fingerprint() == item.fingerprint() }
        val index = when {
            byRecord >= 0 -> byRecord
            byId >= 0 -> byId
            byFingerprint >= 0 -> byFingerprint
            else -> -1
        }
        if (index >= 0) {
            val keep = current[index]
            current[index] = item.copy(
                id = keep.id,
                readingRecordId = item.readingRecordId ?: keep.readingRecordId,
            )
        } else {
            current.add(0, item)
        }
        save(current)
    }

    fun find(recordId: String?, result: CastResult): SavedAIAnalysis? {
        val items = _items.value
        if (!recordId.isNullOrBlank()) {
            items.firstOrNull { it.readingRecordId == recordId }?.let { return it }
        }
        val fp = SavedAIAnalysis.fingerprint(result)
        items.firstOrNull { it.fingerprint() == fp }?.let { return it }
        return items.firstOrNull { matches(it, recordId, result) }
    }

    fun remove(id: String) {
        save(_items.value.filterNot { it.id == id })
    }

    private fun save(items: List<SavedAIAnalysis>) {
        prefs.edit().putString(KEY, json.encodeToString(items)).apply()
        _items.value = items.sortedByDescending { it.updatedAtEpochMs }
    }

    companion object {
        private const val PREFS = "ai.saved.analyses.v1"
        private const val KEY = "items"

        fun matches(item: SavedAIAnalysis, recordId: String?, result: CastResult): Boolean {
            if (!recordId.isNullOrBlank() && item.readingRecordId == recordId) return true
            if (item.fingerprint() == SavedAIAnalysis.fingerprint(result)) return true
            return item.methodRaw == result.method.raw
                && item.primaryNumber == result.primaryNumber
                && item.lines == result.lines.map { it.rawValue }
                && kotlin.math.abs(item.createdAtEpochMs - result.createdAt.toEpochMilli()) < 1000
        }
    }
}
